app/services/work_packages/set_attributes_service.rb

```diff
@@ -51,17 +51,85 @@ def set_static_attributes(attributes)
   end
 
   def set_calculated_attributes(attributes)
+    derivable = derivable_attribute
     if work_package.new_record?
       set_default_attributes(attributes)
     else
       update_dates
     end
     update_duration
+    update_derivable(derivable)
     update_project_dependent_attributes
     reassign_invalid_status_if_type_changed
     set_templated_description
   end
 
+  def derivable_attribute
+    derivable_attribute_by_others_presence || derivable_attribute_by_others_absence
+  end
+
+  # Returns a field derivable by the presence of the two others.
+  #
+  # A field is derivable if it has not been set explicitly while the other two
+  # fields are set. Matching is done in the order :duration, :due_date,
+  # :start_date. The first one to match is returned.
+  def derivable_attribute_by_others_presence
+    if attribute_not_set_in_params?(:duration) && both_present?(:start_date, :due_date)
+      :duration
+    elsif attribute_not_set_in_params?(:due_date) && both_present?(:start_date, :duration)
+      :due_date
+    elsif attribute_not_set_in_params?(:start_date) && both_present?(:due_date, :duration)
+      :start_date
+    end
+  end
+
+  # Returns a field derivable by the absence of one of the two others.
+  #
+  # A field is derivable if it has not been set explicitly while the other two
+  # fields have one set and one nil. Matching is done in the order :duration, :due_date,
+  # :start_date. The first one to match is returned.
+  #
+  # Note: if both other fields are nil, then the field is not derivable
+  def derivable_attribute_by_others_absence
+    if attribute_not_set_in_params?(:duration) && only_one_present?(:start_date, :due_date)
+      :duration
+    elsif attribute_not_set_in_params?(:due_date) && only_one_present?(:start_date, :duration)
+      :due_date
+    elsif attribute_not_set_in_params?(:start_date) && only_one_present?(:due_date, :duration)
+      :start_date
+    end
+  end
+
+  def attribute_not_set_in_params?(field)
+    !params.has_key?(field)
+  end
+
+  def both_present?(*fields)
```

Comment:
Since this method can take any arguments, I think its name rather should be `all_present?(*fields)`.
```suggestion
  def all_present?(*fields)
```

Reply:
arg, I blindly commited it from GitHub web, but the callers are still using the previous method name. I oversaw that. I'll fix it.

Reply:
fixed and force-pushed